File: examples/swift/notes-showcase/Tests/NotesShowcaseTests/StatusBarVMTests.swift
//
// StatusBarVMTests — scenario tests for StatusBarVM.
//
// Ports NotesShowcase.Tests/ViewModels/StatusBarVMTests.cs (C# Avalonia flavor).
// No conformance-ID markers (scenario IDs live in THEME-00x only).
//
// Task-8 note: BindableDerived sidecars are not implemented in Swift yet.
// The two sidecar tests from the C# file are skipped here; they will be
// added in Task 8.
//
import XCTest
import Combine
import VMx
@testable import NotesShowcaseCore

// MARK: - StatusBarVMTests

final class StatusBarVMTests: XCTestCase {

    // MARK: - Shared fixture

    /// All fixtures share a hub so property-changed messages from NotesViewVM /
    /// NoteFormVM reach the StatusBarVM subscription.
    private struct Fixture {
        let hub: MessageHub
        let bar: StatusBarVM
        let notes: NotesViewVM
        let notebooks: NotebooksRootVM
        let form: NoteFormVM
        let repo: InMemoryNoteRepository
    }

    private func build() throws -> Fixture {
        let hub = MessageHub()
        let dispatcher = ImmediateDispatcher.INSTANCE
        let repo = InMemoryNoteRepository(
            seed: SeedData.build(),
            loadAllDelay: 0,
            loadNotesDelay: 0,
            saveNoteDelay: 0
        )
        let notes = try NotesViewVM.builder()
            .name("notes").services(hub: hub, dispatcher: dispatcher)
            .repository(repo).pageSize(5)
            .searchDebounce(.milliseconds(0))
            .build()
        let notebooks = try NotebooksRootVM.builder()
            .name("nbs").services(hub: hub, dispatcher: dispatcher)
            .repository(repo)
            .build()
        let form = try NoteFormVM.builder()
            .name("form").services(hub: hub, dispatcher: dispatcher)
            .repository(repo)
            .build()
        let bar = try StatusBarVM.builder()
            .name("status").services(hub: hub, dispatcher: dispatcher)
            .notesView(notes).notebooks(notebooks).noteForm(form)
            .build()
        try notes.construct()
        try notebooks.construct()
        try form.construct()
        try bar.construct()
        return Fixture(hub: hub, bar: bar, notes: notes, notebooks: notebooks, form: form, repo: repo)
    }

    private func sampleNote(title: String = "Title") -> NoteModel {
        NoteModel(
            id: "n1", notebookId: "nb-reviews", title: title,
            tags: [], body: "", starred: false,
            createdAt: Date(), updatedAt: Date()
        )
    }

    // MARK: - noteCountText

    func testNoteCountText_recomputesWhenNotesViewEmitsPropertyChanged() async throws {
        let f = try build()
        let initial = try f.bar.noteCountText.value
        await f.notes.bindTo(notebookId: "nb-reviews")
        let updated = try f.bar.noteCountText.value
        XCTAssertNotEqual(initial, updated)
        XCTAssertTrue(updated.contains("note"), "Expected 'note' in '\(updated)'")
    }

    // MARK: - starredText

    func testStarredText_reflectsStarredCountInCurrentFilter() async throws {
        let f = try build()
        await f.notes.bindTo(notebookId: "nb-reviews")
        // SeedData: 2 starred notes in nb-reviews.
        XCTAssertEqual("2 starred", try f.bar.starredText.value)
    }

    // MARK: - editingText

    func testEditingText_saysNoSelectionUntilFormIsBound() throws {
        let f = try build()
        XCTAssertEqual("No selection", try f.bar.editingText.value)

        f.form.bindTo(sampleNote(title: "Title"))

        let editing = try f.bar.editingText.value
        XCTAssertTrue(editing.contains("Editing:"), "Expected 'Editing:' in '\(editing)'")
        XCTAssertTrue(editing.contains("Title"),    "Expected 'Title' in '\(editing)'")
    }

    func testEditingText_dirtyMarkerAppearsOnMutation() throws {
        let f = try build()
        f.form.bindTo(sampleNote(title: "Title"))

        XCTAssertFalse((try f.bar.editingText.value).contains(" *"),
                       "Expected no ' *' marker before mutation")

        f.form.draft = f.form.draft.with(title: "Title v2")

        XCTAssertTrue((try f.bar.editingText.value).contains(" *"),
                      "Expected ' *' dirty marker after draft mutation")
    }

    // MARK: - Distinct-until-changed

    func testEqualityGuardNoDuplicateDerivedPropertyEmissionsForSameValue() async throws {
        let f = try build()
        await f.notes.bindTo(notebookId: "nb-reviews")

        var emitCount = 0
        var cancellables = Set<AnyCancellable>()
        f.bar.noteCountText.valueChanged
            .sink { _ in emitCount += 1 }
            .store(in: &cancellables)

        // Same filter = same count = no emission via distinct-until-changed.
        f.notes.filter = nil
        f.notes.filter = nil
        f.notes.showStarredOnly = false

        XCTAssertEqual(0, emitCount,
                       "Expected zero emissions for no-op filter mutations (distinct-until-changed)")
    }
}

File: examples/csharp/avalonia/NotesShowcase.Tests/ViewModels/StatusBarVMTests.cs
using System.Reactive.Concurrency;
using NotesShowcase.Models;
using NotesShowcase.ViewModels;
using VMx.Services;
using Xunit;

namespace NotesShowcase.Tests.ViewModels;

public sealed class StatusBarVMTests
{
    private static (StatusBarVM bar, NotesViewVM notes, NotebooksRootVM nbs, NoteFormVM form,
                    InMemoryNoteRepository repo) Build()
    {
        var hub = new MessageHub();
        var dispatcher = new RxDispatcher(ImmediateScheduler.Instance, ImmediateScheduler.Instance);
        var repo = new InMemoryNoteRepository(
            SeedData.Build(),
            loadAllDelay: TimeSpan.Zero,
            loadNotesDelay: TimeSpan.Zero,
            saveNoteDelay: TimeSpan.Zero);
        var notes = NotesViewVM.Builder()
            .Name("notes").Services(hub, dispatcher).Repository(repo).PageSize(5).Build();
        var nbs = NotebooksRootVM.Builder()
            .Name("nbs").Services(hub, dispatcher).Repository(repo).Build();
        var form = NoteFormVM.Builder()
            .Name("form").Services(hub, dispatcher).Repository(repo).Build();
        var bar = StatusBarVM.Builder()
            .Name("status").Services(hub, dispatcher)
            .NotesView(notes).Notebooks(nbs).NoteForm(form).Build();
        notes.Construct();
        nbs.Construct();
        form.Construct();
        bar.Construct();
        return (bar, notes, nbs, form, repo);
    }

    [Fact]
    public async Task NoteCountText_recomputes_when_NotesView_emits_PropertyChanged()
    {
        var (bar, notes, _, _, _) = Build();
        var initial = bar.NoteCountText.Value;
        await notes.BindToAsync("nb-reviews");
        Assert.NotEqual(initial, bar.NoteCountText.Value);
        Assert.Contains("notes", bar.NoteCountText.Value); // "7 notes"
    }

    [Fact]
    public async Task StarredText_reflects_starred_count_in_current_filter()
    {
        var (bar, notes, _, _, _) = Build();
        await notes.BindToAsync("nb-reviews");
        // 2 starred in nb-reviews per SeedData.
        Assert.Equal("2 starred", bar.StarredText.Value);
    }

    [Fact]
    public void EditingText_says_no_selection_until_form_is_bound()
    {
        var (bar, _, _, form, _) = Build();
        Assert.Equal("No selection", bar.EditingText.Value);
        form.BindTo(new NoteModel(
            "n1", "nb-reviews", "Title", Array.Empty<string>(), "", false,
            DateTimeOffset.UtcNow, DateTimeOffset.UtcNow));
        Assert.Contains("Editing:", bar.EditingText.Value);
        Assert.Contains("Title", bar.EditingText.Value);
    }

    [Fact]
    public void EditingText_dirty_marker_appears_on_mutation()
    {
        var (bar, _, _, form, _) = Build();
        form.BindTo(new NoteModel(
            "n1", "nb-reviews", "Title", Array.Empty<string>(), "", false,
            DateTimeOffset.UtcNow, DateTimeOffset.UtcNow));
        Assert.DoesNotContain(" *", bar.EditingText.Value);
        form.Draft = form.Draft with { Title = "Title v2" };
        Assert.Contains(" *", bar.EditingText.Value);
    }

    [Fact]
    public async Task Equality_guard_no_duplicate_DerivedProperty_emissions_for_same_value()
    {
        var (bar, notes, _, _, _) = Build();
        await notes.BindToAsync("nb-reviews");
        var emits = 0;
        using var sub = bar.NoteCountText.ValueChanged.Subscribe(_ => emits++);
        // Triggering RecomputeFiltered with the same predicate => same count => no emission.
        notes.Filter = null;
        notes.Filter = null;
        notes.ShowStarredOnly = false;
        Assert.Equal(0, emits);
    }

    // ── BindableDerived INPC adapter contract ─────────────────────────────

    [Fact]
    public async Task Bindable_sidecars_raise_PropertyChanged_for_Value_on_recompute()
    {
        var (bar, notes, _, _, _) = Build();
        var observedCount = new System.Collections.Generic.List<string?>();
        var observedStarred = new System.Collections.Generic.List<string?>();
        bar.NoteCountTextBindable.PropertyChanged += (_, e) => observedCount.Add(e.PropertyName);
        bar.StarredTextBindable.PropertyChanged += (_, e) => observedStarred.Add(e.PropertyName);

        await notes.BindToAsync("nb-reviews");

        Assert.NotEmpty(observedCount);
        Assert.All(observedCount, n => Assert.Equal("Value", n));
        Assert.NotEmpty(observedStarred);
        Assert.Equal(bar.NoteCountText.Value, bar.NoteCountTextBindable.Value);
        Assert.Equal(bar.StarredText.Value, bar.StarredTextBindable.Value);
    }

    [Fact]
    public void EditingTextBindable_PropertyChanged_fires_when_form_draft_mutates()
    {
        var (bar, _, _, form, _) = Build();
        var observed = new System.Collections.Generic.List<string?>();
        bar.EditingTextBindable.PropertyChanged += (_, e) => observed.Add(e.PropertyName);

        form.BindTo(new NoteModel(
            "n1", "nb-reviews", "Hello", Array.Empty<string>(), "", false,
            DateTimeOffset.UtcNow, DateTimeOffset.UtcNow));
        form.Title = "World";

        Assert.NotEmpty(observed);
        Assert.Equal(bar.EditingText.Value, bar.EditingTextBindable.Value);
    }
}
